### DISCUSSION TITLE
EntityLinker's loss is huge

### DISCUSSION THREAD

**Discussion by amirj** [explosion/spaCy#8011] [Help: Coding & Implementations] [answered]
I'm training an entity linker model:

```python
for epoch in range(epoches):
    with self.nlp.select_pipes(enable=["entity_linker"]):
        optimizer = self.nlp.resume_training()
        random.shuffle(self.train_set)
        # increasing batch sizes
        batches = minibatch(self.train_set, size=compounding(4.0, 32.0, 1.001))
        losses = {}
        for batch in batches:
            self.nlp.update(batch, drop=0.2, losses=losses, sgd=optimizer)
        logger.info(f"epoch#{epoch+1} losses: {losses}")
    # evaluate on dev set
    score = self.nlp.evaluate(self.dev_set)
    print(score)

```

This is the output for the first epoch:

```
epoch#1 losses: {'entity_linker': 256.69330079441465}

{'token_acc': 1.0, 'token_p': 1.0, 'token_r': 1.0, 'token_f': 1.0, 'tag_acc': None, 'morph_acc': None, 'morph_per_feat': None, 'lemma_acc': None, 'pos_acc': None, 'sents_p': 0.989967692569291, 'sents_r': 0.9901360544217687, 'sents_f': 0.9900518663378964, 'dep_uas': None, 'dep_las': None, 'dep_las_per_type': None, 'ents_p': 0.09771202477206262, 'ents_r': 0.7994370161857847, 'ents_f': 0.17413964896144707, 'ents_per_type': {'ENTITY': {'p': 0.09771202477206262, 'r': 0.7994370161857847, 'f': 0.17413964896144707}}, 'nel_score': 0.4961801367108968, 'nel_score_desc': 'micro F', 'nel_micro_p': 0.6345560507370587, 'nel_micro_r': 0.40735035211267606, 'nel_micro_f': 0.4961801367108968, 'nel_macro_p': 0.6345560507370587, 'nel_macro_r': 0.40735035211267606, 'nel_macro_f': 0.4961801367108968, 'nel_f_per_type': {'ENTITY': {'p': 0.6345560507370587, 'r': 0.40735035211267606, 'f': 0.4961801367108968}}, 'speed': 1355.7736340872466}

```

While the Precision/Recall metrics for entity linker seems reasonable (`'nel_micro_p': 0.63, 'nel_micro_r': 0.40`) but the training loss is huge (`256.69`). Do you have any idea why it does happen?

I just noticed that the EntityRecognizer has very low precision (`'ents_p': 0.097`) and high recall (`'ents_r': 0.799`). The main reason is that I'm using a "general" NER model that detects a variety of entities while doing entity linking for just a very small subset of entity types. Though I'm not sure how it would effects the loss value.

**Comment by svlandeg** [comment, marked as answer]
I wouldn't fixate on the `loss` value too much. Its implementation is component-specific, and it's really difficult to assign meaning to the numeric values. During training, you do want to check that the loss is going down. (and I have seen MUCH higher loss values than 256 ;-)) 

In general your NEL performance metrics seem quite good, so I wouldn't worry about that!

I would worry about the performance of the NER though, because that seems very low. Perhaps you need to train a custom NER model for your challenge/dataset?
